File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

 const cable = consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log('online');
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
    console.log('offline');
  },

  received(data) {

    console.log(data);


//============== IF THE USER IS RECEIVING THE CALL ==============================================
    if (data['step'] === 'receiving the call'){
      var sender_first_name = data['sender_first_name'];
      var session_id = data['session_id'];
      var session_id_modal = document.getElementById('session_id');
      session_id_modal.innerHTML = session_id;
      var sender_name_modal = document.getElementById('sender_name');
      sender_name_modal.innerHTML = sender_first_name;
      // Display the receiver notification modal
      $('#receiver-notif-modal').modal('show');

    }

// ============ BROADCASTING THE SESSION TO THE RECIPIENT.=====================================================
    if (data['step'] === 'Broadcasting session to the recipient') {
      console.log('Broadcasting the session to the recipient');
      // Initialize the session
      var session = OT.initSession(data['apikey'], data['session_id']);
      console.log(session);

      // Initialize the publisher for the recipient
      var publisherProperties = {insertMode: "append", width: '100%', height: '100%'};
      var publisher = OT.initPublisher('publisher', publisherProperties, function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log("Receiver publisher initialized.");
        }
      });
      $('#session-modal').modal("show");

      // Detect when new streams are created and subscribe to them.
      session.on("streamCreated", function (event) {
        console.log("New stream in the session");
        console.log("New stream in the session: " + event.stream);
        var subscriberProperties = {insertMode: 'append', width: '100%', height: '100%'};
        session.subscribe(event.stream, 'subscriber', subscriberProperties, function(error) {
          if (error) {
            console.log(error);
          } else {
            console.log('Receiver subscriber added.');
          }
        });
      });

      // Connect to the session
      // If the connection is successful, publish an audio-video stream.
      session.connect(data['token'], function(error) {
        if (error) {
          console.log("Error connecting: MBEA", error.name, error.message);
        } else {
          console.log("Connected to the session.");
          session.publish(publisher, function(error) {
            if (error) {
              console.log(error);
            } else {
              console.log("The receiver is publishing a stream");
            }
          });
        }
      });

    }

// ============ BROADCASTING THE SESSION TO THE SENDER.=====================================================
    if (data['step'] === 'Broadcasting session to the sender') {
      console.log('Broadcasting the session to the sender');
      // Initialize the session
      var session = OT.initSession(data['apikey'], data['session_id']);
      console.log(session);
      $('#sender-notif-modal').modal("hide");
      // Initialize the publisher for the sender
      var publisherProperties = {insertMode: "append", width: '100%', height: '100%'};
      var publisher = OT.initPublisher('publisher', publisherProperties, function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log("Sender publisher initialized.");
        }
      });
      $('#session-modal').modal("show");
      // Detect when new streams are created and subscribe to them.
      session.on("streamCreated", function (event) {
        console.log("New stream in the session");
        console.log("New stream in the session: " + event.stream);
        var subscriberProperties = {insertMode: 'append', width: '100%', height: '100%'};
        session.subscribe(event.stream, 'subscriber', subscriberProperties, function(error) {
          if (error) {
            console.log(error);
          } else {
            console.log('Sender subscriber added.');
          }
        });
      });
      // Connect to the session
      // If the connection is successful, publish an audio-video stream.
      session.connect(data['token'], function(error) {
        if (error) {
          console.log("Error connecting: MBEA", error.name, error.message);
        } else {
          console.log("Connected to the session.");
          session.publish(publisher, function(error) {
            if (error) {
              console.log(error);
            } else {
              console.log("The sender is publishing a stream");
            }
          });
        }
      });


    }

  },



  call(recipient_id, recipient_name) {
    console.log(`Hello from the call method: ${recipient_id}. You are calling ${recipient_name}`);
    return this.perform('call', {
            recipient_id: recipient_id,
            recipient_name: recipient_name
    });
  },

  answer(session_id) {
    console.log(`Hello from the answer method: ${session_id}`);
    return this.perform('answer', {
            session_id: session_id
    });

  }
});

export default cable;

File: app/channels/room_channel.rb
require "opentok"

class RoomChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    stream_from "room_#{current_user.id}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
    stop_all_streams
  end

  def call(data)
    recipient_id = data['recipient_id']
    recipient_name = data['recipient_name']
    @session = opentokSession
    session_id = @session.session_id
    broadcast_notif_to_recipient(recipient_id, session_id)


  end

  def answer(data)
    session_id = data["session_id"]
    sender_id = data["sender_id"]
    broadcast_session_to_recipient(session_id)
    broadcast_session_to_sender(session_id, sender_id)
  end


  private


  # Brodcast the session to the recipient
  def broadcast_session_to_recipient(session_id)
    token = create_token(session_id)
    ActionCable.server.broadcast(
      "room_#{current_user.id}",
      apikey: api_key,
      session_id: session_id,
      token: token,
      step: 'Broadcasting session to the recipient'
    )
  end

  def broadcast_session_to_sender(session_id, sender_id)
    token = create_token(session_id)
    ActionCable.server.broadcast(
      "room_#{sender_id}",
      apikey: api_key,
      session_id: session_id,
      token: token,
      step: 'Broadcasting session to the sender'
    )
  end

  def broadcast_notif_to_recipient(recipient_id, session_id)
    ActionCable.server.broadcast(
      "room_#{recipient_id}",
      sender_first_name: current_user.full_name,
      sender_id: current_user.id,
      session_id: session_id,
      step: 'receiving the call'
    )
  end

  def opentokSession
    create_session
  end

  def api_key
    Rails.application.credentials.opentok[:api_key]
  end

  def secret_key
    Rails.application.credentials.opentok[:secret_key]
  end

  def opentok
    OpenTok::OpenTok.new api_key, secret_key
  end

  def create_session
    opentok.create_session :media_mode => :routed
  end

  def create_token(session_id)
    opentok.generate_token(session_id)
  end



end
